### PR TITLE
BUG: fix NumPy bridge for itk::Image with PixelType itk::Vector

### PR DESCRIPTION
**Pull request by SimonRit**
@thewtex These changes fixed the issue described on [discourse](https://discourse.itk.org/t/numpy-bridge-with-image-vectors/1422/2).

### CODE REVIEW

**Comment by thewtex** [comment]
Thank, @SimonRit !

I see that we do have Fortran-order tests, but the errors are not propagating, as you mention. We are setting the Python exception state; however, CPython [expects calls to return `NULL` or similar](https://docs.python.org/3.7/c-api/exceptions.html) to examine the exception state (note that this is C exception handling.) But `_GetImageViewFromArray` is returning `const typename PyBuffer<TImage>::OutputImagePointer`... I expect we need to tell SWIG to return NULL or something similar...

Could you please also create a pull request for this branch against `InsightSoftwareConsortium:python-builds`, i.e. the `python-builds` branch? This will trigger Python builds. We are working on improving this workflow, but that is the process for now.

**Comment by SimonRit** [comment]
Done, PR #213.

**Comment by thewtex** [comment]
The [macOS Python build](https://open.cdash.org/index.php?project=Insight&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=NumPyBridgeForImageVectors) is failing the `PythonExtrasTest` with the following

```
Traceback (most recent call last):
  File "extras.py", line 179, in <module>
    image = itk.GetImageFromArray(arr, is_vector=True)
  File "/Users/vsts/agent/2.142.1/work/1/s-build/Wrapping/Generators/Python/itkExtras.py", line 286, in GetImageFromArray
    return _GetImageFromArray(arr, "GetImageFromArray", is_vector)
  File "/Users/vsts/agent/2.142.1/work/1/s-build/Wrapping/Generators/Python/itkExtras.py", line 281, in _GetImageFromArray
    return templatedFunction(arr, is_vector)
  File "/Users/vsts/agent/2.142.1/work/1/s-build/lib/itkPyBufferPython.py", line 3717, in GetImageFromArray
    image._ndarr = array_copy
AttributeError: 'NoneType' object has no attribute '_ndarr'
```

**Comment by SimonRit** [comment]
I admit I did not run the tests, sorry. Something is not right in the CI, the problem should be seen on any platform. In my opinion, this is the expected behavior given the test and the current status of the code. The test might have passed before but I don't think the result was the one expected. If I print the type and shape of arr before [line 179 of extras.py](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Wrapping/Generators/Python/Tests/extras.py#L179), I get (256, 256, 3) and uint8. Since we don't specify the image type, it automatically detects the pixel type as itk.UC and an error message is raised by the new commit [here](https://github.com/InsightSoftwareConsortium/ITK/pull/209/commits/0678b6fc7953a8e831032a6e1b976c77b39e8bb4#diff-88eccc9fc72de83d50d8b9c8441c0deeR112) (but not properly displayed yet).

I see two options to fix this:
- explicitly call the correct type in the test for vector images. For [line 179 of extras.py](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Wrapping/Generators/Python/Tests/extras.py#L179), it would be
```itk.PyBuffer[itk.Image[itk.RGBPixel[itk.UC],2]].GetImageFromArray(arr, is_vector=True)```
This is how I used it.
- guess the correct image type for vectors in [itkExtras.py](https://github.com/InsightSoftwareConsortium/ITK/blob/00f8e256c441c52df493b0c488249a894ba808e6/Wrapping/Generators/Python/itkExtras.py#L268). I tried to do this but this is a bit difficult since we need to handle itk::Vector or RGBPixel.

What do you think? Any preference or other suggestion?

**Comment by thewtex** [comment]
> Something is not right in the CI, the problem should be seen on any platform.

Perhaps NumPy is not available to run the tests on all platforms. This PR is created to add NumPy to the testing environments: https://github.com/InsightSoftwareConsortium/ITK/pull/224

**Comment by thewtex** [comment]
> I see two options to fix this:
> 
>     * explicitly call the correct type in the test for vector images. For [line 179 of extras.py](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Wrapping/Generators/Python/Tests/extras.py#L179), it would be
>       `itk.PyBuffer[itk.Image[itk.RGBPixel[itk.UC],2]].GetImageFromArray(arr, is_vector=True)`
>       This is how I used it.
> 
>     * guess the correct image type for vectors in [itkExtras.py](https://github.com/InsightSoftwareConsortium/ITK/blob/00f8e256c441c52df493b0c488249a894ba808e6/Wrapping/Generators/Python/itkExtras.py#L268). I tried to do this but this is a bit difficult since we need to handle itk::Vector or RGBPixel.
> 
> 
> What do you think? Any preference or other suggestion?

Yes, good thoughts.

We want a "reasonable" default if `is_vector` is `True`. `itk.RGBPixel` if the component type is `itk.UC` and `itk.Vector` otherwise sounds good.

**Comment by SimonRit** [comment]
@thewtex I have amended the patch to (try to) do that. Note the additional patch for improving ImageFileReader with images such as itk::Image<itk::Vector<float, 5>, 3>.

File: Wrapping/Generators/Python/itkExtras.py [hunk 1]
```diff
@@ -274,9 +274,20 @@ def _GetImageFromArray(arr, function, is_vector):
...
         Dimension = arr.ndim - 1
+        if arr.flags['C_CONTIGUOUS']:
+            VectorDimension = arr.shape[-1]
+        else:
+            VectorDimension = arr.shape[0]
+        if PixelType == itk.UC:
+            if VectorDimension == 3:
+                ImageType = itk.Image[ itk.RGBPixel[itk.UC], Dimension ]
+            elif VectorDimension == 4:
+                ImageType = itk.Image[ itk.RGBAPixel[itk.UC], Dimension ]
```
Comment (thewtex) [inline Wrapping/Generators/Python/itkExtras.py]: Nice :+1:

**Comment by thewtex** [comment]
Well done, thanks @SimonRit !

**Pull request merged**